modules/openapi-generator/src/main/resources/python/api_client.mustache

```diff
@@ -640,6 +642,60 @@ class ApiClient(object):
 
         instance = klass(**kwargs)
```

Comment:
Hi @rienafairefr,

I did some further investigation on the issues regarding (multi-level) inheritance.
I hope the following helps you:

1. First of all with your PR we should also apply the following changes on `PythonClientCodegen.java`.
   I suppose this should be correct since the Python client code does not support real inheritance at Python level, but it "_copies_" the _attributes_ to the child classes.

```diff
diff --git a/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientCodegen.java b/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientCodegen.java
index aa6e8ad015..6fc0066264 100644
--- a/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientCodegen.java
+++ b/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientCodegen.java
@@ -57,7 +57,7 @@ public class PythonClientCodegen extends DefaultCodegen implements CodegenConfig
         // at the moment
         importMapping.clear();

-        supportsInheritance = true;
+        //supportsInheritance = true;
         modelPackage = "models";
         apiPackage = "api";
         outputFolder = "generated-code" + File.separatorChar + "python";
```
2. Secondly, when the "_base_" classes (when using `allOf`) or the "_selection_" classes (when using `oneOf`/`anyOf`) have a `discriminator`, we should try that one first. If that one fails, we can use your "_implicit_" hierarchy code.

   The current code especially fails when:

   1. we have a `responseBody` which returns a `BaseClass`
   1. with an object structure as below

   In that case the `if hierarchy['allOf']:` will return an instance of the `BaseClass` instead of an instance of `SomeChild`.

```yaml
schemas:
  RemoteId:
    properties:
      id:
        type: int
  BaseClass:
    discriminator:
      propertyName: implType
    allOf:
    - $ref: RemoteId
    - properties:
        implType:
          type: string
  SomeChild:
    allOf:
    - $ref: BaseClass
    - properties:
        childData:
          type: string
```

(_I would liked to make the suggestion on multiple lines (till the end of the file), but I did not succeed in it, sorry for that_ 😕)
```suggestion
        instance = klass(**kwargs)

        if hasattr(instance, 'get_real_child_model'):
            klass_name = instance.get_real_child_model(data)
            if klass_name:
                instance = self.__deserialize(data, klass_name)

        if hasattr(instance, 'composed_hierarchy'):
            hierarchy = instance.composed_hierarchy
            if sum([1 if v else 0 for v in hierarchy.values()]) > 1:
                # ignore implicit AllOf
                return instance
            def error_msg(composed):
                return "Failed to parse `{0}` as {1} ({2} {3})" \
                    .format(data, klass,
                            composed,
                            ' '.join(hierarchy[composed]))
            if hierarchy['allOf']:
                matches = []
                for sub_klass in hierarchy['allOf']:
                    try:
                        if sub_klass in self.__deserializer_track:
                            # seen it before, break the recursion
                            matches.append(None)
                        else:
                            matches.append(self.__deserialize(data, sub_klass))
                    except:  # noqa: E722
                        pass
                if len(matches) == len(hierarchy['allOf']):
                    self.__deserializer_track = []
                    return instance
                # not all matched -> error
                raise rest.ApiException(status=0, reason=error_msg('allOf'))
            if hierarchy['anyOf']:
                for sub_klass in hierarchy['anyOf']:
                    try:
                        # if at least one of the sub-class matches
                        # terminate with an instance of this class
                        return self.__deserialize(data, sub_klass)
                    except:  # noqa: E722
                        pass
                # none matched -> error
                raise rest.ApiException(status=0, reason=error_msg('anyOf'))
            if hierarchy['oneOf']:
                matches = []
                for sub_klass in hierarchy['oneOf']:
                    try:
                        matches.append(self.__deserialize(data, sub_klass))
                    except:  # noqa: E722
                        pass
                if len(matches) == 1:
                    return matches[0]
                # none matched, or more than one matched -> error
                raise rest.ApiException(status=0, reason=error_msg('oneOf'))

        return instance
```

Reply:
Thanks for the comment, yeah I guess this goes to show that the idea of handling the allOf/oneOf/anyOf at a single level inside a model is bound to fail, because some information about how to deserialize a given file resides not only in the model's, but also in the model's children and parent, we have to go deeper somehow, but not too deeper otherwise we become stuck in recursion. Not sure how to proceed

Reply:
What information is stored in the model parent?